pkg/network/tracer/tracer_linux_test.go

```diff
@@ -1669,3 +1683,386 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				server := NewTCPServerOnAddress(serverAddr, func(c net.Conn) {
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolUnclassified,
+		},
+		{
+			name: "tcp client with sending random data",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				timedContext, cancel := context.WithTimeout(context.Background(), time.Second)
+				c, err := dialer.DialContext(timedContext, "tcp", serverAddr)
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("hello\n"))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				server := NewTCPServerOnAddress(serverAddr, func(c net.Conn) {
+					r := bufio.NewReader(c)
+					input, err := r.ReadBytes(byte('\n'))
+					if err == nil {
+						c.Write(input)
+					}
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolUnknown,
+		},
+		{
+			name: "tcp client with sending HTTP request",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				client := nethttp.Client{
+					Transport: &nethttp.Transport{
+						DialContext: dialer.DialContext,
+					},
+				}
+				resp, err := client.Get("http://" + serverAddr + "/test")
+				require.NoError(t, err)
+				io.Copy(ioutil.Discard, resp.Body)
+				resp.Body.Close()
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				srv := &nethttp.Server{
+					Addr: serverAddr,
+					Handler: nethttp.HandlerFunc(func(w nethttp.ResponseWriter, req *nethttp.Request) {
+						io.Copy(ioutil.Discard, req.Body)
+						w.WriteHeader(200)
+					}),
+					ReadTimeout:  time.Second,
+					WriteTimeout: time.Second,
+				}
+				srv.SetKeepAlivesEnabled(false)
+				go func() {
+					_ = srv.ListenAndServe()
+				}()
+				go func() {
+					<-done
+					fmt.Println("shutting down")
+					srv.Shutdown(context.Background())
+				}()
+			},
+			want: network.ProtocolHTTP,
+		},
+		{
+			// Although it seems weird to test it, it is a real use-case.
+			// Consider a keep alive connection already established between a client and a server.
+			// Then the agent is being started, and able to capture the last response sent, and then the connection
+			// is terminated. Although we won't capture that HTTP traffic, it is still (as far as we know) a HTTP connection.
+			name: "tcp client with sending HTTP response",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				timedContext, cancel := context.WithTimeout(context.Background(), time.Second)
+				c, err := dialer.DialContext(timedContext, "tcp", serverAddr)
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("HTTP/1.1 404 NOT FOUND\r\n"))
```

Comment:
Should the server side be sending this?

Reply:
I documented above the exact scenario

since the client is the one the initiates the connection, I needed it to send the response first

if it makes more sense, I can export it into a dedicate test instead of blending in the current test

Reply:
This doesn't seem right ... wouldn't the server do this call and the client a `c.Read`? In a real world scenario the client wouldn't send a 404, and the server wouldn't see one right?

Reply:
in the real world, the client will never send that

I wanted to mimic a "http keep alive" scenario, when the agent is being deployed while there are http requests being sent to the server
and the first packet it sees it is an http response
exactly like we did in `TestKeepAliveWithIncompleteResponseRegression`